File: frontend/lib/sentry.ts
import type { Breadcrumb, ErrorEvent, EventHint } from '@sentry/nextjs'

const SENSITIVE_HEADERS = new Set([
  'authorization',
  'cookie',
  'x-admin-token',
  'x-user-token',
  'x-company-user-token',
  'x-internal-token',
  // CloudFront が本番の全リクエストへ必ず付ける経路証明トークン(#1407)。
  // 漏れると ALB 直叩きで CloudFront 経由と誤認させ、詐称XFFを署名させられる。
  'x-origin-token',
  // httpContextIntegration が必ず載せる。遷移元のクエリ（＝トークン）が入る
  'referer',
])

/**
 * URL からクエリ文字列とフラグメントを落とす。
 *
 * クエリにワンタイムトークンやユーザー情報を載せている画面が実在する:
 *   /verify-email?token=...            メール認証トークン
 *   /company-portal/setup?token=...    企業担当者の招待トークン
 *   /auth/callback?user=...            Base64 のユーザー情報
 * これらのページでJSエラーが1件起きるだけで、そのまま Sentry へ送られてしまう。
 */
function stripUrlSecrets(url: string): string {
  const cut = url.search(/[?#]/)
  return cut === -1 ? url : url.slice(0, cut)
}

/**
 * Sentry 送信前に認証ヘッダー・Cookie・ボディ・URLのクエリを落とす（#619 / #1185）。
 * 履歴書・チャット本文などの個人情報混入を防ぐ。
 */
export function scrubSentryEvent(event: ErrorEvent, _hint?: EventHint): ErrorEvent | null {
  if (event.request) {
    delete event.request.data
    delete event.request.cookies
    if (event.request.headers) {
      for (const key of Object.keys(event.request.headers)) {
        // Referer にも遷移元のクエリ（＝トークン）が載る
        if (SENSITIVE_HEADERS.has(key.toLowerCase())) {
          delete event.request.headers[key]
        }
      }
    }
    if (event.request.query_string) {
      event.request.query_string = ''
    }
    // httpContextIntegration が window.location.href をそのまま入れるため、
    // query_string を空にするだけでは足りない。
    if (typeof event.request.url === 'string') {
      event.request.url = stripUrlSecrets(event.request.url)
    }
  }
  return event
}

/**
 * パンくず（fetch / navigation / console）からもクエリを落とす。
 * パンくずは beforeSend の対象外なので、ここで潰さないと素通りする。
 */
export function scrubSentryBreadcrumb(breadcrumb: Breadcrumb): Breadcrumb | null {
  if (typeof breadcrumb.data?.url === 'string') {
    breadcrumb.data.url = stripUrlSecrets(breadcrumb.data.url)
  }
  if (typeof breadcrumb.data?.from === 'string') {
    breadcrumb.data.from = stripUrlSecrets(breadcrumb.data.from)
  }
  if (typeof breadcrumb.data?.to === 'string') {
    breadcrumb.data.to = stripUrlSecrets(breadcrumb.data.to)
  }
  // console のパンくずは引数をそのまま持つ。何が載るか読めないので丸ごと落とす。
  if (breadcrumb.category === 'console') return null
  return breadcrumb
}

export function sentrySharedOptions() {
  const dsn =
    process.env.SENTRY_DSN?.trim() ||
    process.env.NEXT_PUBLIC_SENTRY_DSN?.trim() ||
    ''
  if (!dsn) return null

  return {
    dsn,
    // NEXT_PUBLIC_ が無いとクライアント側では undefined になり、
    // staging と production のイベントが Sentry 上で区別できない。
    environment:
      process.env.APP_ENV ||
      process.env.NEXT_PUBLIC_APP_ENV ||
      process.env.NODE_ENV ||
      'development',
    release: process.env.SENTRY_RELEASE || process.env.NEXT_PUBLIC_SENTRY_RELEASE || undefined,
    sendDefaultPii: false,
    tracesSampleRate: 0,
    beforeSend: scrubSentryEvent,
    beforeBreadcrumb: scrubSentryBreadcrumb,
  }
}

File: Backend/internal/observability/sentry.go
// Package observability はエラートラッキング等の運用計装を提供する（#619 / #1185）。
package observability

import (
	"log/slog"
	"os"
	"strings"
	"time"

	"github.com/getsentry/sentry-go"
)

// InitSentry は SENTRY_DSN が設定されているときだけ Sentry を初期化する。
// DSN 未設定時は no-op（ローカル開発で依存しない）。
// 戻り値の flush はプロセス終了前に呼ぶこと。
func InitSentry() (flush func(), enabled bool) {
	dsn := strings.TrimSpace(os.Getenv("SENTRY_DSN"))
	if dsn == "" {
		return func() {}, false
	}

	env := strings.TrimSpace(os.Getenv("APP_ENV"))
	if env == "" {
		env = "development"
	}
	release := strings.TrimSpace(os.Getenv("SENTRY_RELEASE"))

	err := sentry.Init(sentry.ClientOptions{
		Dsn:              dsn,
		Environment:      env,
		Release:          release,
		SendDefaultPII:   false,
		BeforeSend:       scrubEvent,
		TracesSampleRate: 0, // トレースは別Issue。エラー通知のみ。
	})
	if err != nil {
		slog.Error("sentry init failed", "error", err)
		return func() {}, false
	}
	slog.Info("sentry enabled", "environment", env, "release", release)
	return func() {
		sentry.Flush(2 * time.Second)
	}, true
}

// sensitiveHeaders は Sentry へ送ってはいけないリクエストヘッダー（小文字）。
// frontend の `frontend/lib/sentry.ts` の SENSITIVE_HEADERS と同じ集合にすること。
// 片方だけ直して他方を忘れる事故が実際に起きたため、TestSensitiveHeaders_フロントと同期
// が両者の突き合わせで落ちる。
var sensitiveHeaders = map[string]bool{
	"authorization":        true,
	"cookie":               true,
	"x-admin-token":        true,
	"x-user-token":         true,
	"x-company-user-token": true,
	"x-internal-token":     true,
	// CloudFront が本番の全オリジンリクエストへ付ける経路証明トークン(#1407)。
	// distribution の aliases は `*.shukatsu-ai.jp` を含むため、api ドメインを
	// CloudFront 経由で叩けば backend にもこのヘッダー付きで届く。漏れると
	// 公開ALBへの frontend 直叩きで詐称 X-Forwarded-For を署名させられる。
	"x-origin-token": true,
	// 遷移元のクエリ（＝ワンタイムトークン）がそのまま載る。
	"referer": true,
}

// scrubEvent は Sentry 送信前に認証ヘッダー・Cookie・リクエストボディを落とす。
// 履歴書・チャット本文などの個人情報が混入しないようにする（#619）。
func scrubEvent(event *sentry.Event, _ *sentry.EventHint) *sentry.Event {
	if event == nil {
		return nil
	}
	if event.Request != nil {
		event.Request.Cookies = ""
		event.Request.Data = ""
		for key := range event.Request.Headers {
			if sensitiveHeaders[strings.ToLower(key)] {
				delete(event.Request.Headers, key)
			}
		}
		event.Request.QueryString = ""
	}

	// 例外メッセージは素通りする。err.Error() に外部レスポンス本文や AI 出力を
	// そのまま埋めている箇所が実在する（resume_review.go の "rag review failed: %s" など）。
	// 長い本文ほど個人情報を含みやすいので、頭だけ残して切る。
	for i := range event.Exception {
		event.Exception[i].Value = truncateForSentry(event.Exception[i].Value)
	}
	event.Message = truncateForSentry(event.Message)
	return event
}

// maxSentryMessageRunes は例外メッセージとして送る上限。
// 原因の特定には先頭で足り、それ以上は本文の持ち出しになりやすい。
const maxSentryMessageRunes = 300

func truncateForSentry(msg string) string {
	runes := []rune(msg)
	if len(runes) <= maxSentryMessageRunes {
		return msg
	}
	return string(runes[:maxSentryMessageRunes]) + "…(truncated)"
}
